frontend/src/metabase/components/FieldValuesWidget/FieldValuesWidget.jsx

```diff
@@ -68,405 +74,364 @@ function mapStateToProps(state, { fields = [] }) {
+    setLoadingState("LOADING");
+    setOptions([]);
 
-    let options = [];
-    let valuesMode = this.state.valuesMode;
+    let newOptions = [];
+    let newValuesMode = valuesMode;
     try {
-      if (canUseDashboardEndpoints(this.props.dashboard)) {
-        const { values, has_more_values } =
-          await this.fetchDashboardParameterValues(query);
-        options = values;
-        valuesMode = has_more_values ? "search" : valuesMode;
-      } else if (canUseCardEndpoints(this.props.question)) {
-        const { values, has_more_values } = await this.fetchCardParameterValues(
+      if (canUseDashboardEndpoints(dashboard)) {
+        const { values, has_more_values } = await fetchDashboardParameterValues(
           query,
         );
-        options = values;
-        valuesMode = has_more_values ? "search" : valuesMode;
-      } else if (canUseParameterEndpoints(this.props.parameter)) {
-        const { values, has_more_values } = await this.fetchParameterValues(
+        newOptions = values;
+        newValuesMode = has_more_values ? "search" : newValuesMode;
+      } else if (canUseCardEndpoints(question)) {
+        const { values, has_more_values } = await fetchCardParameterValues(
           query,
         );
-        options = values;
-        valuesMode = has_more_values ? "search" : valuesMode;
+        newOptions = values;
+        newValuesMode = has_more_values ? "search" : newValuesMode;
+      } else if (canUseParameterEndpoints(parameter)) {
+        const { values, has_more_values } = await fetchParameterValues(query);
+        newOptions = values;
+        newValuesMode = has_more_values ? "search" : newValuesMode;
       } else {
-        options = await this.fetchFieldValues(query);
-        const {
-          parameter,
-          fields,
-          disableSearch,
-          disablePKRemappingForSearch,
-        } = this.props;
-        valuesMode = getValuesMode({
+        newOptions = await fetchFieldValues(query);
+
+        newValuesMode = getValuesMode({
           parameter,
           fields,
           disableSearch,
           disablePKRemappingForSearch,
         });
       }
     } finally {
-      this.updateRemappings(options);
-      this.setState({
-        loadingState: "LOADED",
-        options,
-        valuesMode,
-      });
+      updateRemappings(newOptions);
+
+      setOptions(newOptions);
+      setLoadingState("LOADED");
+      setValuesMode(newValuesMode);
     }
-  }
+  };
 
-  fetchFieldValues = async query => {
+  const fetchFieldValues = async query => {
     if (query == null) {
-      const { fetchFieldValues } = this.props;
-      await Promise.all(
-        getNonVirtualFields(this.props.fields).map(field =>
-          fetchFieldValues(field.id),
-        ),
+      const nonVirtualFields = getNonVirtualFields(fields);
+
+      const results = await Promise.all(
+        nonVirtualFields.map(field => fetchFieldValuesProp({ id: field.id })),
       );
```

Comment:
I had to tweak this a bit. The previous logic relied on mutations to `this.props.fields` in the middle of this function. Instead, this relies entirely on the return value from the entity loader.